src/coreclr/nativeaot/System.Private.CoreLib/src/System/CrashInfo.cs

```diff
@@ -0,0 +1,402 @@
+            if (!WriteExceptionHelper(key, exception, maxMessageSize, maxNumberStackFrames, maxMethodNameSize))
+            {
+                _currentBufferIndex = savedIndex;
+                return false;
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Adds the exception info to the JSON buffer
+        /// </summary>
+        /// <param name="key">the UTF8 name of the block</param>
+        /// <param name="exception">exception build triage block from</param>
+        /// <param name="maxMessageSize">limits the size of the exception message strings</param>
+        /// <param name="maxNumberStackFrames">limits the number of stack frames written to the triage buffer</param>
+        /// <param name="maxMethodNameSize">limits the size of the stack frame method name strings</param>
+        /// <returns>true - success, false - out of triage buffer space</returns>
+        private bool WriteExceptionHelper(ReadOnlySpan<byte> key, Exception exception, int maxMessageSize, int maxNumberStackFrames, int maxMethodNameSize)
+        {
+            if (!OpenValue(key, '{'))
+                return false;
+
+            if (!WriteHexValue("address"u8, (ulong)Unsafe.AsPointer(ref exception)))
+                return false;
+
+            if (!WriteHexValue("hr"u8, exception.HResult))
+                return false;
+
+            if (!WriteValue("message"u8, exception.Message, maxMessageSize))
+                return false;
+
+            // Exception type names are not truncated because the full name is important to bucketing and usually not that long
+            if (!WriteValue("type"u8, exception.GetType().ToString()))
+                return false;
+
+            StackFrame[] stackFrames = new StackTrace(exception).GetFrames();
+            if (stackFrames.Length > 0)
+            {
+                if (!OpenValue("stack"u8, '['))
+                    return false;
+
+                int count = 0;
+                foreach (StackFrame frame in stackFrames)
+                {
+                    // Check if the stack frame limit has been hit
+                    if (++count > maxNumberStackFrames)
+                        break;
+
+                    if (!WriteStackFrame(frame, maxMethodNameSize))
+                        return false;
+                }
+
+                CloseValue(']');
+            }
+
+            AggregateException? aggregate = exception as AggregateException;
+            if (aggregate is not null || exception.InnerException is not null)
+            {
+                if (!OpenValue("inner"u8, '['))
+                    return false;
+
+                // Write as many inner exceptions that will fit
+                if (aggregate is not null)
+                {
+                    foreach (Exception ex in aggregate.InnerExceptions)
+                    {
+                        if (!WriteExceptionWithFallback(default, ex, maxMessageSize, maxNumberStackFrames, maxMethodNameSize))
```

Comment:
Is recursion OK here?

Reply:
Yes, but maybe I should limit the number of levels.

Reply:
I'm not going to add any limit on recursion for now because if the inner exception doesn't fit in the buffer it is skipped with no failure.